### PR TITLE
Scatter gun patches to improve entity tracking and position tracking.

### PR DESCRIPTION
**Pull request by gabizou**
A port of @Aikar's patches for Forge 1.12.2 (they can be back ported however seen fit or by request).

https://github.com/PaperMC/Paper/blob/fd1bd5223a461b6d98280bb8f2d67280a30dd24a/Spigot-Server-Patches/0306-Mark-chunk-dirty-anytime-entities-change-to-guarante.patch
https://github.com/PaperMC/Paper/blob/fd1bd5223a461b6d98280bb8f2d67280a30dd24a/Spigot-Server-Patches/0315-Always-process-chunk-registration-after-moving.patch
https://github.com/PaperMC/Paper/blob/fd1bd5223a461b6d98280bb8f2d67280a30dd24a/Spigot-Server-Patches/0335-Ensure-chunks-are-always-loaded-on-hard-position-set.patch
https://github.com/PaperMC/Paper/blob/fd1bd5223a461b6d98280bb8f2d67280a30dd24a/Spigot-Server-Patches/0378-Sync-Player-Position-to-Vehicles.patch

In short, these patches fix several issues with entity and player position tracking in relation to chunks loaded and chunk entity lists themselves. Provided these patches, they together fix a few exploits that otherwise are likely to be fixed in future versions of Minecraft. As an example, someone [has already reported some of their exploits have finally been fixed by these changes from Paper](https://www.reddit.com/r/2b2t/comments/9iedck/the_great_speed_exploit_of_2018_patched/).

I've done my best to keep the patches as small as possible for future maintenance and easy portability to previous versions.

### CODE REVIEW

**Comment by williewillus** [comment]
are there vanilla tickets for any of these? I'd add them so we can track when to remove them

File: patches/minecraft/net/minecraft/entity/Entity.java.patch [hunk 1]
```diff
@@ -9,18 +9,19 @@
...
-@@ -177,6 +177,10 @@
+@@ -177,6 +177,11 @@
      private boolean field_184237_aG;
      private final double[] field_191505_aI;
      private long field_191506_aJ;
 +    /**
 +     * Setting this to true will prevent the world from calling {@link #onUpdate()} for this entity.
 +     */
 +    public boolean updateBlocked;
++    public boolean isWorldActiveInChunk = false;
```
Comment (aikar) [inline patches/minecraft/net/minecraft/entity/Entity.java.patch]: This should just be named isInWorld; it has no relation to a chunk. An Entity can exist in a World but not exist in any Chunk (I don't like that this is possible, but sadly Mojang did this)

File: patches/minecraft/net/minecraft/entity/Entity.java.patch [hunk 2]
```diff
@@ -80,15 +97,23 @@
              }
              else
              {
-@@ -1707,6 +1720,7 @@
+@@ -1382,6 +1398,7 @@
+             this.field_70126_B -= 360.0F;
+         }
+ 
++        this.field_70170_p.func_72964_e((int) Math.floor(this.field_70165_t) >> 4, (int) Math.floor(this.field_70161_v) >> 4); // Forge - ensure chunk loads occur on position changes
```
Comment (aikar) [inline patches/minecraft/net/minecraft/entity/Entity.java.patch]: Comment would be more accurate to say Ensure target chunk is loaded.

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 3]
```diff
@@ -237,15 +237,31 @@
              this.func_72964_e(i, j).func_76612_a(p_72838_1_);
              this.field_72996_f.add(p_72838_1_);
              this.func_72923_a(p_72838_1_);
-@@ -1227,6 +1291,7 @@
+@@ -1153,6 +1217,7 @@
+         {
+             ((IWorldEventListener)this.field_73021_x.get(i)).func_72703_a(p_72923_1_);
+         }
++        p_72923_1_.isWorldActiveInChunk = true;
```
Comment (aikar) [inline patches/minecraft/net/minecraft/world/World.java.patch]: Forge comments?

File: patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch [hunk 4]
```diff
@@ -115,15 +115,30 @@
...
++        this.func_76630_e(); // Forge - ensure chunks are marked to save after an entity add
+     }
+ 
+     public void func_76622_b(Entity p_76622_1_)
+@@ -762,7 +762,7 @@
+             p_76608_2_ = this.field_76645_j.length - 1;
+         }
+ 
+-        this.field_76645_j[p_76608_2_].remove(p_76608_1_);
++        if (!this.field_76645_j[p_76608_2_].remove(p_76608_1_)) this.func_76630_e(); // Forge - ensure chunks are marked to save after entity removals only
```
Comment (aikar) [inline patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch]: don't think you need the if statement here, but I think you have it inverted anyways? This is only marking if it failed to remove the entity.

**Review by aikar — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch (hunk 4 above)
Comment (gabizou) [reply patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch]: From the paper source, ti looks like it only called to update the field if and only if the entity removal was successful (yes, there's still a logical flip that needs to be corrected).

**Comment by gabizou** [comment]
> are there vanilla tickets for any of these? I'd add them so we can track when to remove them

@aikar might have a better idea as far as which ones are directly related. I'll do a search later today for them throughout several discussions.

File: patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch (hunk 4 above)
Comment (aikar) [reply patches/minecraft/net/minecraft/world/chunk/Chunk.java.patch]: Yeah but that was because we have other operations being done in that method increasing/decreasing counts, that we want to ensure only runs when something actually changed state.

**Comment by aikar** [comment]
> Just curious, if an entity is constantly moving, how to prevent that entity from loading too many chunks?

nothing does. If they aren't on a vehicle, this really isn't any difference.

I hope no-one encourages 'use exploits to reduce how many chunks you load on our server while traveling please'

**Comment by aikar** [comment]
Original MC ticket but it's marked as private: https://bugs.mojang.com/browse/MC-136995

**Comment by gabizou** [comment]
Updated the commit with issues fixed. Made the field private and centered it inside the Forge added section of Entity.

**Comment by aikar** [comment]
something I just thought about that we can prob improve on this - boats have 2 passengers. It's likely better if we update all of the vehicles passengers and not just the controller.

**Comment by liach** [comment]
the world class has a blank line patch in imports.

File: patches/minecraft/net/minecraft/entity/Entity.java.patch [hunk 5]
```diff
@@ -46,7 +46,23 @@
...
+         this.field_70161_v = p_70107_5_;
++        if (this.isAddedToWorld) this.field_70170_p.func_72866_a(this, false); // Forge - Process chunk registration after moving.
+         float f = this.field_70130_N / 2.0F;
+         float f1 = this.field_70131_O;
+         this.func_174826_a(new AxisAlignedBB(p_70107_1_ - (double)f, p_70107_3_, p_70107_5_ - (double)f, p_70107_1_ + (double)f, p_70107_3_ + (double)f1, p_70107_5_ + (double)f));
+@@ -995,6 +1008,7 @@
+         this.field_70165_t = (axisalignedbb.field_72340_a + axisalignedbb.field_72336_d) / 2.0D;
+         this.field_70163_u = axisalignedbb.field_72338_b;
+         this.field_70161_v = (axisalignedbb.field_72339_c + axisalignedbb.field_72334_f) / 2.0D;
++        if (this.isAddedToWorld) this.field_70170_p.func_72866_a(this, false); // Forge - Process chunk registration after moving.
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/Entity.java.patch:61]: Will this cause loading of a ton of chunks if the entity moves too fast (e.g. a minecart)?

File: patches/minecraft/net/minecraft/world/World.java.patch [hunk 6]
```diff
@@ -1,6 +1,14 @@
 --- ../src-base/minecraft/net/minecraft/world/World.java
 +++ ../src-work/minecraft/net/minecraft/world/World.java
-@@ -63,8 +63,15 @@
+@@ -11,6 +11,7 @@
+ import java.util.Random;
+ import java.util.UUID;
+ import java.util.function.Supplier;
++
```
Comment (liach) [inline patches/minecraft/net/minecraft/world/World.java.patch:7]: No blank line in imports

File: patches/minecraft/net/minecraft/entity/Entity.java.patch [hunk 7]
```diff
@@ -229,12 +253,47 @@
...
++     * {@link World#onEntityRemoved(Entity)}.
++     *
++     * @return True if this entity is being tracked by a world
++     */
++    public final boolean isAddedToWorld() { return this.isAddedToWorld; }
++
++    /**
++     * Called after the entity has been added to the world's
++     * ticking list. Can be overriden, but needs to call super
++     * to prevent MC-136995.
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/Entity.java.patch:281]: Preferably mention pr number, too.

**Review by liach — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/entity/Entity.java.patch (hunk 5 above)
Comment (aikar) [reply patches/minecraft/net/minecraft/entity/Entity.java.patch:61]: @liach doesn't matter if it does, needs to happen. But no, ideally the chunks would of already been loaded around the player. this is more targeted around teleporting.

But no matter what, if that chunk is unloaded, it needs to load, otherwise the entity is desynced with the current location which is what lead to all this.

File: patches/minecraft/net/minecraft/entity/Entity.java.patch [hunk 8]
```diff
@@ -80,15 +96,23 @@
              }
              else
              {
-@@ -1707,6 +1720,7 @@
+@@ -1382,6 +1397,7 @@
+             this.field_70126_B -= 360.0F;
+         }
+ 
++        this.field_70170_p.func_72964_e((int) Math.floor(this.field_70165_t) >> 4, (int) Math.floor(this.field_70161_v) >> 4); // Forge - ensure target chunk is loaded.
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/Entity.java.patch:103]: Sorry, I mean this `getChunk` one instead of that `updateEntityWithOptionalForce` method might lead to excessive chunk loading with fast minecarts, etc.

File: patches/minecraft/net/minecraft/entity/Entity.java.patch [hunk 9]
```diff
@@ -229,12 +253,47 @@
...
++     * @return True if this entity is being tracked by a world
++     */
++    public final boolean isAddedToWorld() { return this.isAddedToWorld; }
++
++    /**
++     * Called after the entity has been added to the world's
++     * ticking list. Can be overriden, but needs to call super
++     * to prevent MC-136995.
++     */
++    public void onAddedToWorld() {
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/Entity.java.patch:283]: You can annotate with `@javax.annotations.OverridingMethodsMustInvokeSuper`

File: patches/minecraft/net/minecraft/entity/Entity.java.patch (hunk 8 above)
Comment (aikar) [reply patches/minecraft/net/minecraft/entity/Entity.java.patch:103]: The getChunk line was specifically placed here to reduce it to only player controlled movement changes, and not the setPosition method that is called for literally everything (which would hurt performance)

This line is very important to happen BEFORE the .setPosition called right below it. the setPosition following this will then call the updateEntityWithOptionalForce method.

Both are occurring, but the behavior of updateEntityWithOptionalForce changes if the target chunk is not loaded, so we force load it here if it was not (teleporting)
